src/main/kotlin/com/varabyte/kobweb/intellij/colors/KobwebColorProvider.kt

```diff
@@ -0,0 +1,222 @@
+                }
+
+                callee.isKobwebColorFunction("hsl(h: Float, s: Float, l: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(3) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+
+                callee.isKobwebColorFunction("hsla(h: Float, s: Float, l: Float, a: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(4) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+            }
+        }
+
+        else -> null
+    }
+
+    return if (currentDepth <= MAX_SEARCH_DEPTH) {
+        nextElement?.let { traceColor(it, currentDepth + 1) }
+    } else null
+}
+
+/**
+ * Pretty much emulates what CTRL-clicking on a reference would do
+ */
+private fun KtSimpleNameExpression.locateSource(): PsiElement? {
+    val resolvedMainReference = this.mainReference.resolve() ?: return null
+    val sourceDescriptor =
+        PsiNavigationSupport.getInstance().getDescriptor(resolvedMainReference) as? OpenFileDescriptor ?: return null
+    val sourcePsiFile = sourceDescriptor.file.toPsiFile(this.project) ?: return null
+    val sourceElementIdentifier = PsiUtilBase.getElementAtOffset(sourcePsiFile, sourceDescriptor.offset)
+    val sourceElement = sourceElementIdentifier.parent
+
+    return sourceElement
+}
+
+/**
+ * Uses the kotlin analysis api, as it can parse the constants much smarter.
+ * It can parse decimal, hex and binary automatically for example.
+ */
+private inline fun <reified Evaluated, reified Mapped> Collection<KtValueArgument>.evaluateArguments(
+    argCount: Int,
+    evaluatedValueMapper: (Evaluated) -> Mapped
+): Array<Mapped>? {
+    val constantExpressions = this.mapNotNull { it.getArgumentExpression() as? KtConstantExpression }
+
+    val evaluatedArguments = constantExpressions.mapNotNull {
+        analyze(it.containingKtFile) {
+            it.evaluate(KtConstantEvaluationMode.CONSTANT_LIKE_EXPRESSION_EVALUATION)?.value as? Evaluated
+        }
+    }
+
+    return if (evaluatedArguments.size != this.size || evaluatedArguments.size != argCount) null
+    else evaluatedArguments.map(evaluatedValueMapper).toTypedArray()
+}
+
+private inline fun <reified Evaluated> Collection<KtValueArgument>.evaluateArguments(argCount: Int) =
+    evaluateArguments<Evaluated, Evaluated>(argCount) { it }
+
+private fun KtNamedFunction.isKobwebColorFunction(vararg functionSignatures: String): Boolean {
+    val actualFqName = this.kotlinFqName?.asString() ?: return false
+    val actualParameters = this.valueParameterList?.text ?: return false
+    val actual = actualFqName + actualParameters
+
+    return functionSignatures.any { functionSignature ->
+        val expected = "$KOBWEB_COLOR_COMPANION_FQ_NAME.$functionSignature"
+        expected == actual
+    }
+}
+
+private fun safeRgbColor(r: Int, g: Int, b: Int) =
```

Comment:
Optional:

```
/**
 * Returns a color assuming all the input arguments are in valid range, or null otherwise.
 */
```

Reply:
Is there a way to shadow the kdoc to other functions or should I just add it to a single one?

Reply:
Ah actually, this is a pretty big limitation of kdoc. There's no way, and it sucks.

Three options:

* Don't document / let the code be self documenting.
   * `tryCreateColor` is probably obvious as is.
* Document each three, with a very short document. 
  * Note that the docs will be slightly different for the `(r, g, b)`, and `(rgb)` versions, because what's considered invalid is different in all of them. Even hsl has different constraints for what's in range.
* Document a region (since these are private and not public)

Would look something like this:

```
# region create colors safely

// The following methods are guaranteed not to throw, which would
// otherwise happen if you tried to construct a color with some out-
// of-bounds color.

# endregion
```

Believe it or not, I think of all these cases, I'd vote for "don't document". I think in retrospect the code is really clear.

Reply:
Yeah I'd say so too. Creating a color isnt rocket science